pkg/operators/wasm/api/datasource.go

```diff
@@ -0,0 +1,250 @@
+		sub.cb.(PacketFunc)(DataSource(ds), Packet(data))
+	}
+}
+
+type (
+	Packet       uint32
+	DataSource   uint32
+	Field        uint32
+	Data         uint32
+	DataArray    uint32
+	PacketSingle uint32
+	PacketArray  uint32
+)
+
+func GetDataSource(name string) (DataSource, error) {
+	ret := getDataSource(uint64(stringToBufPtr(name)))
+	runtime.KeepAlive(name)
+	if ret == 0 {
+		return 0, fmt.Errorf("datasource %s not found", name)
+	}
+	return DataSource(ret), nil
+}
+
+func NewDataSource(name string) (DataSource, error) {
+	ret := newDataSource(uint64(stringToBufPtr(name)))
+	runtime.KeepAlive(name)
+	if ret == 0 {
+		return 0, fmt.Errorf("creating datasource %q", name)
+	}
+	return DataSource(ret), nil
+}
+
+func (ds DataSource) Subscribe(cb DataFunc, priority uint32) error {
+	dsSubscriptionCtr++
+	dsSubcriptions[dsSubscriptionCtr] = subscription{typ: subscriptionTypeData, cb: cb}
+	ret := dataSourceSubscribe(uint32(ds), uint32(subscriptionTypeData), priority, dsSubscriptionCtr)
+	if ret != 0 {
+		return fmt.Errorf("subscribing to datasource")
+	}
+	return nil
+}
+
+func (ds DataSource) SubscribeArray(cb ArrayFunc, priority uint32) error {
+	dsSubscriptionCtr++
+	dsSubcriptions[dsSubscriptionCtr] = subscription{typ: subscriptionTypeArray, cb: cb}
+	ret := dataSourceSubscribe(uint32(ds), uint32(subscriptionTypeArray), priority, dsSubscriptionCtr)
+	if ret != 0 {
+		return fmt.Errorf("subscribing to datasource")
+	}
+	return nil
+}
+
+func (ds DataSource) SubscribePacket(cb PacketFunc, priority uint32) error {
+	dsSubscriptionCtr++
+	dsSubcriptions[dsSubscriptionCtr] = subscription{typ: subscriptionTypePacket, cb: cb}
+	ret := dataSourceSubscribe(uint32(ds), uint32(subscriptionTypePacket), priority, dsSubscriptionCtr)
+	if ret != 0 {
+		return fmt.Errorf("subscribing to datasource")
+	}
+	return nil
+}
+
+func (ds DataSource) NewPacketSingle() (PacketSingle, error) {
+	ret := dataSourceNewPacketSingle(uint32(ds))
+	if ret == 0 {
+		return 0, errors.New("creating packet")
+	}
+	return PacketSingle(ret), nil
+}
+
+func (ds DataSource) NewPacketArray() (PacketArray, error) {
+	ret := dataSourceNewPacketArray(uint32(ds))
+	if ret == 0 {
+		return 0, errors.New("creating packet")
+	}
+	return PacketArray(ret), nil
+}
+
+func (ds DataSource) EmitAndRelease(packet Packet) error {
+	ret := dataSourceEmitAndRelease(uint32(ds), uint32(packet))
+	if ret != 0 {
+		return fmt.Errorf("emitting data")
+	}
+	return nil
+}
+
+func (ds DataSource) Release(packet Packet) {
+	dataSourceRelease(uint32(ds), uint32(packet))
+}
+
+func (ds DataSource) GetField(name string) (Field, error) {
+	ret := dataSourceGetField(uint32(ds), uint64(stringToBufPtr(name)))
+	runtime.KeepAlive(name)
+	if ret == 0 {
+		return 0, fmt.Errorf("field %q not found", name)
+	}
+	return Field(ret), nil
+}
+
+func (ds DataSource) AddField(name string, kind FieldKind) (Field, error) {
+	ret := dataSourceAddField(uint32(ds), uint64(stringToBufPtr(name)), uint32(kind))
+	runtime.KeepAlive(name)
```

Comment:
I don't think it'llt work as we need `name` to be kept alive until the bufPtr pointing to it is used, i.e. after `dataSourceAddField` returns. 

If we move this to `stringToBufPtr`, it could happen that the garbage collection frees it before calling `dataSourceAddField`, this will make the whole thing fail as the host will try to read something invalid.